File: languages/cpp/projects/01-foundations/main.cpp
#include <iostream>
#include <limits>
#include <string>
#include <vector>

struct Student {
    std::string name;
    int score;
};

int main() {
    int count = 0;
    std::cout << "How many students? ";
    std::cin >> count;

    if (count <= 0) {
        std::cout << "Please enter a positive number.\n";
        return 0;
    }

    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');

    std::vector<Student> students;
    students.reserve(static_cast<std::size_t>(count));

    for (int i = 0; i < count; ++i) {
        Student student;
        std::cout << "Student name " << (i + 1) << ": ";
        std::getline(std::cin, student.name);

        std::cout << "Score for " << student.name << " (0-100): ";
        std::cin >> student.score;
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');

        students.push_back(student);
    }

    int minScore = students[0].score;
    int maxScore = students[0].score;
    int sum = 0;

    std::cout << "\nStudents:\n";
    for (const Student& student : students) {
        std::cout << "- " << student.name << ": " << student.score << '\n';
        sum += student.score;
        if (student.score < minScore) {
            minScore = student.score;
        }
        if (student.score > maxScore) {
            maxScore = student.score;
        }
    }

    const double average = static_cast<double>(sum) / students.size();
    std::cout << "Average: " << average << '\n';
    std::cout << "Minimum: " << minScore << '\n';
    std::cout << "Maximum: " << maxScore << '\n';

    return 0;
}

File: languages/go/projects/01-foundations/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Student struct {
	Name  string
	Score int
}

func readPositiveCount(scanner *bufio.Scanner) int {
	for {
		fmt.Print("How many students? ")
		if !scanner.Scan() {
			continue
		}

		count, err := strconv.Atoi(strings.TrimSpace(scanner.Text()))
		if err != nil {
			fmt.Println("Invalid number. Try again.")
			continue
		}
		if count <= 0 {
			fmt.Println("Please enter a positive number.")
			continue
		}

		return count
	}
}

func readScore(scanner *bufio.Scanner, studentName string) int {
	for {
		fmt.Printf("Score for %s (0-100): ", studentName)
		if !scanner.Scan() {
			continue
		}

		score, err := strconv.Atoi(strings.TrimSpace(scanner.Text()))
		if err != nil {
			fmt.Println("Invalid score. Enter a number from 0 to 100.")
			continue
		}
		if score < 0 || score > 100 {
			fmt.Println("Score out of range. Enter a value from 0 to 100.")
			continue
		}

		return score
	}
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	count := readPositiveCount(scanner)
	students := make([]Student, 0, count)

	for index := 0; index < count; index++ {
		fmt.Printf("Student name %d: ", index+1)
		scanner.Scan()
		name := scanner.Text()
		score := readScore(scanner, name)
		students = append(students, Student{Name: name, Score: score})
	}

	minScore := students[0].Score
	maxScore := students[0].Score
	sum := 0

	fmt.Println()
	fmt.Println("Students:")
	for _, student := range students {
		fmt.Printf("- %s: %d\n", student.Name, student.Score)
		sum += student.Score
		if student.Score < minScore {
			minScore = student.Score
		}
		if student.Score > maxScore {
			maxScore = student.Score
		}
	}

	average := float64(sum) / float64(len(students))
	fmt.Printf("Average: %v\n", average)
	fmt.Printf("Minimum: %d\n", minScore)
	fmt.Printf("Maximum: %d\n", maxScore)
}
